python/hsfs/engine/python.py

```diff
@@ -948,6 +1054,10 @@ def save_dataframe(
         online_write_options: Dict[str, Any],
         validation_id: Optional[int] = None,
     ) -> Optional[job.Job]:
+        if feature_group.time_travel_format == "DELTA":
+            self._check_duplicate_records(dataframe, feature_group)
```

Comment:
Not sure why you put it there. You already do some conversion stuff in the delta engine before you write. Maybe it would be better there?

Reply:
because if fg is online enabled, it does not go through delta engine.